### PR TITLE
build tao bridge to work with tf serving

### PR DESCRIPTION
**Pull request by Orion34-lanbo**
Only test for cpu version tf serving now, current code can support for tensorflow>=2.5.0.

### CODE REVIEW

File: tao/BUILD [hunk 1]
```diff
@@ -1,15 +1,19 @@
 load("@com_google_protobuf//:protobuf.bzl", "cc_proto_library")
 load("@local_config_tf//:build_defs.bzl", "tf_copts")
 load("@local_config_cuda//cuda:build_defs.bzl", "cuda_library", "if_cuda")
+
+# NOTE(lanbo.llb): if_dcu when building for tf serving is just a dummy impl
+load("@local_config_rocm//rocm:build_defs.bzl", "if_dcu", "if_rocm")
```
Comment (qiuxiafei) [inline tao/BUILD:6]: Why DCU rules has connection with tf serving options?

File: tao/BUILD [hunk 2]
```diff
@@ -91,30 +102,43 @@ cc_library(
...
+    deps = [
+        ":version_header",
+        "@org_tao_compiler//tensorflow/compiler/mlir/xla/ral:ral_bridge",
+    ] + if_tf_serving(
+        [
+            "@org_tensorflow//tensorflow/core:lib",
+            "@org_tensorflow//tensorflow/core:framework",
+            "@org_tensorflow//tensorflow/core/common_runtime:core_cpu",
+            "@org_tensorflow//tensorflow/core/platform:stream_executor_no_cuda",
+            "@org_tensorflow//tensorflow/stream_executor:stream_executor_impl",
```
Comment (qiuxiafei) [inline tao/BUILD:134]: Does this `org_tensorflow` come from `tf_community` ? If so, `tao_bridge` is not built against expected tf_serving version?

File: third_party/bazel/tf_source/no_tf_serving_workspace1.bzl.tpl [hunk 3]
```diff
@@ -0,0 +1,10 @@
+def _repositories():
+    native.local_repository(
+        name = "org_tensorflow",
+        path = "../tf_community/",
+    )
+
+def workspace():
+    _repositories()
+
+tf_source_workspace1 = workspace
```
Comment (qiuxiafei) [inline third_party/bazel/tf_source/no_tf_serving_workspace1.bzl.tpl:10]: nit: maybe `no_tf_xxx` is not a good enough name ?

File: third_party/bazel/common.bzl [hunk 4]
```diff
@@ -345,3 +346,150 @@ def get_env_bool_value_str(repository_ctx, env_name):
...
+    """
+    if not url.startswith("https://"):
+        return [url]
+    return [
+        "https://storage.googleapis.com/mirror.tensorflow.org/%s" % url[8:],
+        url,
+    ]
+
+def _get_env_var(ctx, name):
+    if name in ctx.os.environ:
```
Comment (qiuxiafei) [inline third_party/bazel/common.bzl:393]: I think `ctx.os.environ`, which is a `dict`, has a `get` method.

**Comment by qiuxiafei** [comment]
> Does this `org_tensorflow` come from `tf_community` ? If so, `tao_bridge` is not built against expected tf_serving version? 

I've figured it out, I think.

File: tao/.bazelrc [hunk 5]
```diff
@@ -13,11 +13,14 @@ build:avx_linux --copt=-mavx
 build:avx_linux --host_copt=-mavx
 
 build:release_cpu_linux --config=avx_linux
-build:release_cpu_linux --crosstool_top="@ubuntu18.04-gcc7_manylinux2010-cuda11.2-cudnn8.1-tensorrt7.2_config_cuda//crosstool:toolchain"
+# build:release_cpu_linux --crosstool_top="@ubuntu18.04-gcc7_manylinux2010-cuda11.0-cudnn8-tensorrt7.1_config_cuda//crosstool:toolchain"
+build:release_cpu_linux --crosstool_top="@ubuntu16.04-py3-gcc7_manylinux2010-cuda10.0-cudnn7-tensorrt5.1_config_cuda//crosstool:toolchain"
+
```
[the remark this answers is not in the document]
Comment (Orion34-lanbo) [reply tao/.bazelrc:18]: `@ubuntu18.04-gcc7_manylinux2010-cuda11.0-cudnn8-tensorrt7.1_config_cuda//crosstool:toolchain` does not exist in tf-2.4.x, will fail for 2.4.x build

File: tao/BUILD (hunk 1 above)
Comment (Orion34-lanbo) [reply tao/BUILD:6]: Currently not supported or even tested for dcu, just make a dummy `if_dcu` here to make this `BUILD` work without futher change about `rocm` build

File: tao/BUILD (hunk 2 above)
Comment (Orion34-lanbo) [reply tao/BUILD:134]: Tensorflow source code from `tf_community` or release version tgz from `tf_http_archive` can both be `org_tensorflow`, when building against `tf_serving`, the one from `tf_community` will not be used.

File: third_party/bazel/tf_source/no_tf_serving_workspace1.bzl.tpl (hunk 3 above)
Comment (Orion34-lanbo) [reply third_party/bazel/tf_source/no_tf_serving_workspace1.bzl.tpl:10]: Maybe `tf_community_workspace.bzl` ?

Comment (qiuxiafei) [reply third_party/bazel/tf_source/no_tf_serving_workspace1.bzl.tpl:10]: looks better :P

File: tao/BUILD (hunk 1 above)
Comment (qiuxiafei) [reply tao/BUILD:6]: Even more confused when I see another `if_dcu` on line 19...

**Pull request closed without merging**